cats/cats-redis/src/main/java/com/netflix/spinnaker/cats/redis/cluster/ClusteredSortAgentScheduler.java

```diff
@@ -0,0 +1,204 @@
+  }
+
+  private boolean acquireAgent(String type) {
+    try (Jedis jedis = jedisSource.getJedis()) {
+      return jedis.eval(atomicSwapSetsScript(score(TIMEOUT)),
+          Arrays.asList(new String[]{WAITING_SET, WORKING_SET}),
+          Collections.singletonList(type)) != null;
+    }
+  }
+
+  private boolean releaseAgent(String type) {
+    try (Jedis jedis = jedisSource.getJedis()) {
+      return jedis.eval(atomicSwapSetsScript(score(INTERVAL)),
+          Arrays.asList(new String[]{WORKING_SET, WAITING_SET}),
+          Collections.singletonList(type)) != null;
+    }
+  }
+
+  private void saturatePool() {
+    try (Jedis jedis = jedisSource.getJedis()) {
+      // First cull threads in the WORKING set that have been there too long (TIMEOUT time).
+      Set<String> oldKeys = jedis.zrangeByScore(WORKING_SET, "-inf", score(NOW));
+      for (String key : oldKeys) {
+        // Ignore result, since if this agent was released between now and the above jedis call, our work was done
+        // for us.
+        releaseAgent(key);
+      }
+
+      // Now look for agents that have been in the queue for at least INTERVAL time.
+      List<String> keys = Arrays.asList((String[]) jedis.zrangeByScore(WAITING_SET, "-inf", score(NOW)).toArray());
+      Set<AgentWorker> workers = new HashSet<>();
+
+      // Loop until we either run out of threads to use, or agents (which are keys) to run.
+      while ((parallelism == -1 || runningAgents != parallelism) && !keys.isEmpty()) {
+        String agent = keys.remove(0);
+
+        AgentWorker worker = agents.getOrDefault(agent, null);
+        if (worker != null && acquireAgent(agent)) {
+          workers.add(worker);
+          runningAgents++;
+        }
+      }
+
+      for (AgentWorker worker : workers) {
+        agentWorkPool.submit(worker);
+      }
+    }
+  }
+
+  private static class AgentWorker implements Runnable {
+    private final Agent agent;
+    private final AgentExecution agentExecution;
+    private final ExecutionInstrumentation executionInstrumentation;
+    private final ClusteredSortAgentScheduler scheduler;
+
+    AgentWorker(Agent agent, AgentExecution agentExecution, ExecutionInstrumentation executionInstrumentation, ClusteredSortAgentScheduler scheduler) {
+      this.agent = agent;
+      this.agentExecution = agentExecution;
+      this.executionInstrumentation = executionInstrumentation;
+      this.scheduler = scheduler;
+    }
+
+    @Override
+    public void run() {
+      try {
+        executionInstrumentation.executionStarted(agent);
+        long startTime = System.nanoTime();
+        agentExecution.executeAgent(agent);
+        executionInstrumentation.executionCompleted(agent, TimeUnit.NANOSECONDS.toMillis(System.nanoTime() - startTime));
+      } catch (Throwable cause) {
+        executionInstrumentation.executionFailed(agent, cause);
+      } finally {
+        scheduler.releaseAgent(agent.getAgentType());
+        scheduler.runningAgents--;
```

Comment:
nevermind! but maybe it would make sense to put the decrement into the releaseAgent method